src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs

```diff
@@ -373,13 +385,45 @@ private byte[] SignUsingOffsetRsa(byte[] bytes, int offset, int count)
 
         private bool VerifyRsa(byte[] bytes, byte[] signature)
         {
+#if NET6_0_OR_GREATER
+            return VerifyUsingSpan(isRSA: true, bytes, signature);
+#else
             return RSA.VerifyHash(HashAlgorithm.ComputeHash(bytes), signature, HashAlgorithmName, RSASignaturePadding);
+#endif
         }
 
         private bool VerifyUsingOffsetRsa(byte[] bytes, int offset, int count, byte[] signature)
         {
+#if NET6_0_OR_GREATER
+            return VerifyUsingSpan(isRSA: true, bytes.AsSpan(offset, count), signature);
+#else
             return RSA.VerifyHash(HashAlgorithm.ComputeHash(bytes, offset, count), signature, HashAlgorithmName, RSASignaturePadding);
```

Comment:
Since you already have the HashAlgorithmName, why not just call VerifyData and let the framework do the hashing for you (which won't create a temporary array)?

Reply:
> and let the framework do the hashing for you (which won't create a temporary array)?

This doesn't appear to be the case. When I step into it, I get:

https://github.com/dotnet/runtime/blob/9926d609dab5929941b7924a04521ea2c2df27f3/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSA.cs#L661-L662

Which allocates the `byte[]`.

Here is my test code (on windows):

```C#
using System.Security.Cryptography;

RSA rSA = RSA.Create();

byte[] data = [0x1, 0x2, 0x3, 0x4];

byte[] signature = rSA.SignData(data, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

for (int i = 0; i < 1_000; i++)
{
    rSA.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
}
```

![image](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/assets/8291187/3131e79c-2613-472f-8bf9-fa29b9484c71)

Reply:
Should I submit a similar PR to RSA and ECDsa in dotnet/runtime?

Reply:
It looks like SignData(ReadOnlySpan) uses span-hashing, but SignData(byte[]) uses array-hashing; because we never override `public virtual bool VerifyData(byte[], int, int, byte[], HashAlgorithmName, RSASignaturePadding)`.  So the change in dotnet/runtime to make everyone happier is probably to override that method on all N of our RSA types.  (RSACryptoServiceProvider, RSACng, RSAOpenSsl, RSASecurityTransforms, RSAAndroid, and maybe up to double that from the "RSA.Create()" types).  Which is going to involve a lot of copy/paste of the argument validation.  Bleh.

Here, though, it means
```C#
return rsa.VerifyData(
#if NET5_OR_GREATER
    bytes.AsSpan(offset, count),
#else
    bytes, offset, count,
#endif
    signature,
    HashAlgorithmName,
    RSASignaturePadding);
```

should take your numbers down to 0.

Reply:
RSA is still allocating here:

https://github.com/dotnet/runtime/blob/9926d609dab5929941b7924a04521ea2c2df27f3/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSA.cs#L322

![image](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/assets/8291187/c483a07e-9e32-4884-be30-ae35e0359c0e)

`HashData` returns a `new byte[]`.

Reply:
Weird, since it should have hit the `if (this is IRuntimeAlgorithm)` above and bypassed the allocation routine.  Must've missed a marker interface somewhere, too.

Reply:
In my case, `this` is an `RSABCrypt`, which doesn't implement that interface.

https://github.com/dotnet/runtime/blob/acd31754892ab0431ac2c40038f541ffa7168be7/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSABCrypt.cs#L11